contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,212 @@
+# Device Manager Proposal
+
+  * [Abstract](#abstract)
+  * [Motivation](#motivation)
+  * [Use Cases](#use-cases)
+  * [Objectives](#objectives)
+  * [Proposed Changes](#proposed-changes)
+
+
+_Authors:_
+
+* @RenaudWasTaken - Renaud Gaubert &lt;rgaubert@nvidia.com&gt;
+
+## Abstract
+
+This documents describes a solution to discovering, monitoring and representing
+external devices such as:
+  * GPUs
+  * NICs
+  * FPGAs
+  * InfiniBand
+  * ...
+
+## Motivation
+
+Kubernetes currently supports discovery of CPU and Memory primarily to a
+minimal extent. Very few devices are handled natively by Kubelet.
+
+It is not a sustainable solution to expect every vendor to add their vendor
+specific code inside kubernetes. This approach does not scale and is not portable.
+
+We want a solution for those vendors to be able to advertise their resources to the kubelet
+and monitor them.
+We also want a way for the user to specify which resource their jobs will use and what
+constraints are associated to these resources.
+
+In order to solve this problem it is obvious that we need a plugin system in
+order to have vendors advertise and monitor their resources on behalf of Kubelet.
+
+Additionally, we introduce the concept of ResourceType to be able to select
+resources with constraints in a pod spec.
+
+_GPU Integration Example:_
+  * [Enable "kick the tires" support for Nvidia GPUs in COS](https://github.com/kubernetes/kubernetes/pull/45136)
+  * [Extend experimental support to multiple Nvidia GPUs](https://github.com/kubernetes/kubernetes/pull/42116)
+
+_Kubernetes Meeting Notes On This:_
+  * [Meeting notes](https://docs.google.com/document/d/1Qg42Nmv-QwL4RxicsU2qtZgFKOzANf8fGayw8p3lX6U/edit#)
+  * [Better Abstraction for Compute Resources in Kubernetes](https://docs.google.com/document/d/1666PPUs4Lz56TqKygcy6mXkNazde-vwA7q4e5H92sUc)
+  * [Extensible support for hardware devices in Kubernetes (join kubernetes-dev@googlegroups.com for access)](https://docs.google.com/document/d/1LHeTPx_fWA1PdZkHuALPzYxR0AYXUiiXdo3S0g2VSlo/edit)
+
+## Use Cases
+
+  * I want to use a particular device type (GPU, InfiniBand, FPGA, etc.) in my pod.
+  * I should be able use that device without writing custom Kubernetes code.
+  * I want some mechanism to give priority among pods for particular devices on my node.
+
+## Objectives
+
+1. Create a plugin mechanism which allows discovery and monitoring of devices
+2. Change the API to support RessourceClass and ResourceType
+3. Add support for ResourceType and ResourceType in the scheduler and kubelet
+
+## Non Objectives
+
+1. Solving the runtime is out of scope
+
+## Proposed changes
+
+### API Changes
+#### ResourceType
+
+When discovering the devices, Kubelet will be in charge of advertising those
+resources to the API server.
+
+We will advertise each device returned by the Device Plugin in a new structure
+called the ResourceType.
+It is defined as follows:
+
+```golang
+type ResourceType struct {
```

Comment:
Yeah - That's an issue - tying us to much to PCI/USB devices (where bus, class, vendor, device would work).